### PR TITLE
Fixed google bug 154855417

### LINKED ISSUE #3884: Crash when Google Maps goes down
#### Problem description

It looks like [this crash](https://console.firebase.google.com/u/a0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/71b0f8dfadb8435206e57c3b072dcb92?time=last-ninety-days&sessionId=5ED447A5006F00014CFC40CB8EDAC83C_DNE_0_v2) is occurring when there is an outage on the Google Maps side. 

Google have created a fix and deployed in the new Play Services but have also posted workaround code to prevent the crash on devices with older Play Services: https://issuetracker.google.com/issues/154855417#comment509. We should probably incorporate this in our code.

#### Steps to reproduce the problem

Not really possible as it's an internal Google Maps SDK problem.

#### Expected behavior

The app does not crash if the Google Maps API is having problems.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #3884

#### What has been done to verify that this works as intended?
Not much since we can't reproduce it. I just prepared my changes according to https://issuetracker.google.com/issues/154855417#comment509

#### Why is this the best possible solution? Were any other approaches considered?
I implemented changes according to the explanation provided here (as I said above): https://issuetracker.google.com/issues/154855417#comment509
I adjusted the sample fix for android keeping just the fix for maps v2 and using our shared preferences.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
We are not able to reproduce the issue and also testing doesn't make sense.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java [hunk 1]
```diff
@@ -91,6 +92,9 @@
     @Inject
     public CollectJobCreator collectJobCreator;
 
+    @Inject
+    public GeneralSharedPreferences generalSharedPreferences;
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: This should be using `MetaSharedPreferencesProvider` instead. That way we don't need to use `KEYS_WE_SHOULD_NOT_RESET`.

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java [hunk 2]
```diff
@@ -150,6 +154,7 @@ public void onCreate() {
         singleton = this;
 
         setupDagger();
+        fixGoogleBug154855417();
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java:158]: Probably nice to add a comment with a link to the issue given this is a weird one!

**Review by seadowg — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java:158]: Done.

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Done.

**Review by seadowg — APPROVED** [review]

**Comment by seadowg** [comment]
@grzesiek2010 seeing some weird issues on Circle CI. Could you try rerunning "from start" to see if it fixes it?

**Comment by grzesiek2010** [comment]
> seeing some weird issues on Circle CI. Could you try rerunning "from start" to see if it fixes it?

it's green.

**Pull request merged**